test/core/xds/xds_transport_fake.h

```diff
@@ -0,0 +1,180 @@
+//
+// Copyright 2022 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#ifndef GRPC_CORE_EXT_XDS_XDS_TRANSPORT_FAKE_H
+#define GRPC_CORE_EXT_XDS_XDS_TRANSPORT_FAKE_H
+
+#include <grpc/support/port_platform.h>
+
+#include <deque>
+#include <functional>
+#include <memory>
+#include <string>
+
+#include "absl/status/status.h"
+#include "absl/types/optional.h"
+
+#include "src/core/ext/xds/xds_bootstrap.h"
+#include "src/core/ext/xds/xds_transport.h"
+#include "src/core/lib/gprpp/orphanable.h"
+#include "src/core/lib/gprpp/ref_counted_ptr.h"
+#include "src/core/lib/gprpp/sync.h"
+
+namespace grpc_core {
+
+class FakeXdsTransportFactory : public XdsTransportFactory {
+ private:
+  class FakeXdsTransport;
+
+ public:
+  static constexpr char kAdsMethod[] =
+      "/envoy.service.discovery.v3.AggregatedDiscoveryService/"
+      "StreamAggregatedResources";
+  static constexpr char kAdsV2Method[] =
+      "/envoy.service.discovery.v2.AggregatedDiscoveryService/"
+      "StreamAggregatedResources";
+
+  class FakeStreamingCall : public XdsTransport::StreamingCall {
+   public:
+    FakeStreamingCall(
+        RefCountedPtr<FakeXdsTransport> transport, const char* method,
+        std::unique_ptr<StreamingCall::EventHandler> event_handler)
+        : transport_(std::move(transport)),
+          method_(method),
+          event_handler_(MakeRefCounted<RefCountedEventHandler>(
+              std::move(event_handler))) {}
+
+    void Orphan() override;
+
+    using StreamingCall::Ref;  // Make it public.
+
+    bool HaveMessageFromClient();
+    absl::optional<std::string> WaitForMessageFromClient(
+        absl::Duration timeout);
+
+    void SendMessageToClient(absl::string_view payload);
+    void MaybeSendStatusToClient(absl::Status status);
+
+   private:
+    class RefCountedEventHandler : public RefCounted<RefCountedEventHandler> {
```

Comment:
would it make sense to just make `EventHandler` as a Refcounted class

Reply:
In the production code, it doesn't actually need to be ref-counted; the ref-counting is needed only here in the test code.  I don't think it makes sense to add complexity to the production code for something we need only in the test.